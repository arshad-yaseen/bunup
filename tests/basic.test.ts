import { beforeEach, describe, expect, it } from "vitest";
import {
    cleanProjectDir,
    createProject,
    findFile,
    runBuild,
    validateBuildFiles,
} from "./utils";

describe("bunup basic build", () => {
    beforeEach(() => {
        cleanProjectDir();
    });

    it("should build with ESM format", async () => {
        createProject({
            "index.ts": `export const result = "Hello, world!";`,
        });

        const result = await runBuild({
            entry: "index.ts",
            format: ["esm"],
        });

        expect(result.success).toBe(true);
        expect(result.files.length).toBe(1);

        expect(validateBuildFiles(result, ["index.mjs"])).toBe(true);
    });

    it("should build with CJS format", async () => {
        createProject({
            "index.ts": `export const result = "Hello, world!";`,
        });

        const result = await runBuild({
            entry: "index.ts",
            format: ["cjs"],
        });

        expect(result.success).toBe(true);
        expect(result.files.length).toBe(1);

        expect(validateBuildFiles(result, ["index.js"])).toBe(true);
    });

    it("should build with default format", async () => {
        createProject({
            "index.ts": `export const result = "Hello, world!";`,
        });

        const result = await runBuild({
            entry: "index.ts",
            format: ["iife"],
        });

        expect(result.success).toBe(true);
        expect(result.files.length).toBe(1);

        expect(validateBuildFiles(result, ["index.global.js"])).toBe(true);
    });

    it("should build with multiple formats simultaneously", async () => {
        createProject({
            "index.ts": `export const result = "Hello, world!";`,
        });

        const result = await runBuild({
            entry: "index.ts",
            format: ["esm", "cjs", "iife"],
        });

        expect(result.success).toBe(true);
        expect(result.files.length).toBe(3);

        expect(
            validateBuildFiles(result, [
                "index.mjs",
                "index.js",
                "index.global.js",
            ]),
        ).toBe(true);

        const esmFile = findFile(result, "index", ".mjs");
        expect(esmFile?.content).toMatchSnapshot();

        const cjsFile = findFile(result, "index", ".js");
        expect(cjsFile?.content).toMatchSnapshot();

        const iifeFile = findFile(result, "index", ".global.js");
        expect(iifeFile?.content).toMatchSnapshot();
    });

    it("should output .cjs file when package.json type is module", async () => {
        createProject({
            "index.ts": `export const result = "Hello, world!";`,
            "package.json": `{
                "name": "test",
                "version": "1.0.0",
                "type": "module"
            }`,
        });

        const result = await runBuild({
            entry: "index.ts",
            format: ["cjs"],
        });

        expect(result.success).toBe(true);
        expect(result.files.length).toBe(1);

        expect(validateBuildFiles(result, ["index.cjs"])).toBe(true);
    });
});
