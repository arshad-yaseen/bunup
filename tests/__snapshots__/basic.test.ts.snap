// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`bunup basic build should build with multiple formats simultaneously: 
          "// tests/.project/index.ts
          var result = "Hello, world!";
          export {
            result
          };
          "
         1`] = `
"// tests/.project/index.ts
var result = "Hello, world!";
export {
  result
};
"
`;

exports[`bunup basic build should build with multiple formats simultaneously: 
          "var __defProp = Object.defineProperty;
          var __getOwnPropNames = Object.getOwnPropertyNames;
          var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
          var __hasOwnProp = Object.prototype.hasOwnProperty;
          var __moduleCache = /* @__PURE__ */ new WeakMap;
          var __toCommonJS = (from) => {
            var entry = __moduleCache.get(from), desc;
            if (entry)
              return entry;
            entry = __defProp({}, "__esModule", { value: true });
            if (from && typeof from === "object" || typeof from === "function")
              __getOwnPropNames(from).map((key) => !__hasOwnProp.call(entry, key) && __defProp(entry, key, {
                get: () => from[key],
                enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
              }));
            __moduleCache.set(from, entry);
            return entry;
          };
          var __export = (target, all) => {
            for (var name in all)
              __defProp(target, name, {
                get: all[name],
                enumerable: true,
                configurable: true,
                set: (newValue) => all[name] = () => newValue
              });
          };

          // tests/.project/index.ts
          var exports__ = {};
          __export(exports__, {
            result: () => result
          });
          module.exports = __toCommonJS(exports__);
          var result = "Hello, world!";
          "
         1`] = `
"var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __moduleCache = /* @__PURE__ */ new WeakMap;
var __toCommonJS = (from) => {
  var entry = __moduleCache.get(from), desc;
  if (entry)
    return entry;
  entry = __defProp({}, "__esModule", { value: true });
  if (from && typeof from === "object" || typeof from === "function")
    __getOwnPropNames(from).map((key) => !__hasOwnProp.call(entry, key) && __defProp(entry, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    }));
  __moduleCache.set(from, entry);
  return entry;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {
      get: all[name],
      enumerable: true,
      configurable: true,
      set: (newValue) => all[name] = () => newValue
    });
};

// tests/.project/index.ts
var exports__ = {};
__export(exports__, {
  result: () => result
});
module.exports = __toCommonJS(exports__);
var result = "Hello, world!";
"
`;

exports[`bunup basic build should build with multiple formats simultaneously: 
          "(() => {
            var __defProp = Object.defineProperty;
            var __getOwnPropNames = Object.getOwnPropertyNames;
            var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
            var __hasOwnProp = Object.prototype.hasOwnProperty;
            var __moduleCache = /* @__PURE__ */ new WeakMap;
            var __toCommonJS = (from) => {
              var entry = __moduleCache.get(from), desc;
              if (entry)
                return entry;
              entry = __defProp({}, "__esModule", { value: true });
              if (from && typeof from === "object" || typeof from === "function")
                __getOwnPropNames(from).map((key) => !__hasOwnProp.call(entry, key) && __defProp(entry, key, {
                  get: () => from[key],
                  enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
                }));
              __moduleCache.set(from, entry);
              return entry;
            };
            var __export = (target, all) => {
              for (var name in all)
                __defProp(target, name, {
                  get: all[name],
                  enumerable: true,
                  configurable: true,
                  set: (newValue) => all[name] = () => newValue
                });
            };

            // tests/.project/index.ts
            var exports__ = {};
            __export(exports__, {
              result: () => result
            });
            var result = "Hello, world!";
          })();
          "
         1`] = `
"(() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __moduleCache = /* @__PURE__ */ new WeakMap;
  var __toCommonJS = (from) => {
    var entry = __moduleCache.get(from), desc;
    if (entry)
      return entry;
    entry = __defProp({}, "__esModule", { value: true });
    if (from && typeof from === "object" || typeof from === "function")
      __getOwnPropNames(from).map((key) => !__hasOwnProp.call(entry, key) && __defProp(entry, key, {
        get: () => from[key],
        enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
      }));
    __moduleCache.set(from, entry);
    return entry;
  };
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, {
        get: all[name],
        enumerable: true,
        configurable: true,
        set: (newValue) => all[name] = () => newValue
      });
  };

  // tests/.project/index.ts
  var exports__ = {};
  __export(exports__, {
    result: () => result
  });
  var result = "Hello, world!";
})();
"
`;

exports[`bunup basic build should build with multiple formats simultaneously 1`] = `
"// tests/.project/index.ts
var result = "Hello, world!";
export {
  result
};
"
`;

exports[`bunup basic build should build with multiple formats simultaneously 2`] = `
"var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __moduleCache = /* @__PURE__ */ new WeakMap;
var __toCommonJS = (from) => {
  var entry = __moduleCache.get(from), desc;
  if (entry)
    return entry;
  entry = __defProp({}, "__esModule", { value: true });
  if (from && typeof from === "object" || typeof from === "function")
    __getOwnPropNames(from).map((key) => !__hasOwnProp.call(entry, key) && __defProp(entry, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    }));
  __moduleCache.set(from, entry);
  return entry;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {
      get: all[name],
      enumerable: true,
      configurable: true,
      set: (newValue) => all[name] = () => newValue
    });
};

// tests/.project/index.ts
var exports__ = {};
__export(exports__, {
  result: () => result
});
module.exports = __toCommonJS(exports__);
var result = "Hello, world!";
"
`;

exports[`bunup basic build should build with multiple formats simultaneously 3`] = `
"(() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __moduleCache = /* @__PURE__ */ new WeakMap;
  var __toCommonJS = (from) => {
    var entry = __moduleCache.get(from), desc;
    if (entry)
      return entry;
    entry = __defProp({}, "__esModule", { value: true });
    if (from && typeof from === "object" || typeof from === "function")
      __getOwnPropNames(from).map((key) => !__hasOwnProp.call(entry, key) && __defProp(entry, key, {
        get: () => from[key],
        enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
      }));
    __moduleCache.set(from, entry);
    return entry;
  };
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, {
        get: all[name],
        enumerable: true,
        configurable: true,
        set: (newValue) => all[name] = () => newValue
      });
  };

  // tests/.project/index.ts
  var exports__ = {};
  __export(exports__, {
    result: () => result
  });
  var result = "Hello, world!";
})();
"
`;
